### PR TITLE
[Merged by Bors] - feat: port CategoryTheory.Limits.Shapes.Terminal

### PR DESCRIPTION
**Pull request by mattrobball**
---

- [x] depends on: #2368
[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by semorrison** [comment]
This PR/issue depends on:
* leanprover-community/mathlib4#2368
By **[Dependent Issues](https://github.com/z0al/dependent-issues)** (🤖). Happy coding!

**Comment by semorrison** [comment]
This PR/issue depends on:
* ~~leanprover-community/mathlib4#2368~~
By **[Dependent Issues](https://github.com/z0al/dependent-issues)** (🤖). Happy coding!

File: Mathlib/CategoryTheory/Limits/Shapes/Terminal.lean [hunk 1]
```diff
@@ -31,20 +31,33 @@ namespace CategoryTheory.Limits
...
 @[simps]
 def asEmptyCocone (X : C) : Cocone (Functor.empty.{0} C) :=
-  { x
-    ι := by tidy }
+  { pt := X
+    ι := 
+    { app := by dee
+      naturality := by dee
+    }
+  }
```
Comment (jcommelin) [inline Mathlib/CategoryTheory/Limits/Shapes/Terminal.lean:60]:
```suggestion
      naturality := by dee } }
```

File: Mathlib/CategoryTheory/Limits/Shapes/Terminal.lean [hunk 2]
```diff
@@ -31,20 +31,33 @@ namespace CategoryTheory.Limits
 
 variable {C : Type u₁} [Category.{v₁} C]
 
-attribute [local tidy] tactic.discrete_cases
+-- attribute [local dee] tactic.discrete_cases -- Porting note: no dee
```
Comment (jcommelin) [inline Mathlib/CategoryTheory/Limits/Shapes/Terminal.lean:34]:
```suggestion
-- attribute [local tidy] tactic.discrete_cases -- Porting note: no tidy
```

File: Mathlib/CategoryTheory/Limits/Shapes/Terminal.lean [hunk 3]
```diff
@@ -31,20 +31,33 @@ namespace CategoryTheory.Limits
...
+open Lean Elab Meta Tactic in 
+-- Porting note: trying a macro 
+/-- A local tactic replacing the use of discrete cases in Lean 3 -/
+elab (name := discrete_empty_elim) "dee" : tactic => withMainContext do
```
Comment (jcommelin) [inline Mathlib/CategoryTheory/Limits/Shapes/Terminal.lean:38]: What does `dee` stand for? Please add the explanation to the docstring.

File: Mathlib/CategoryTheory/Limits/Shapes/Terminal.lean [hunk 4]
```diff
@@ -31,20 +31,33 @@ namespace CategoryTheory.Limits
 
 variable {C : Type u₁} [Category.{v₁} C]
 
-attribute [local tidy] tactic.discrete_cases
+-- attribute [local dee] tactic.discrete_cases -- Porting note: no dee
+open Lean Elab Meta Tactic in 
+-- Porting note: trying a macro
```
Comment (jcommelin) [inline Mathlib/CategoryTheory/Limits/Shapes/Terminal.lean:36]:
```suggestion
-- Porting note: adding a local macro to replace tidy.
```

File: Mathlib/CategoryTheory/Limits/Shapes/Terminal.lean (hunk 1 above)
Comment (mattrobball) [reply Mathlib/CategoryTheory/Limits/Shapes/Terminal.lean:60]: Done.

File: Mathlib/CategoryTheory/Limits/Shapes/Terminal.lean (hunk 3 above)
Comment (mattrobball) [reply Mathlib/CategoryTheory/Limits/Shapes/Terminal.lean:40]: I added this to the docstring.

File: Mathlib/CategoryTheory/Limits/Shapes/Terminal.lean (hunk 4 above)
Comment (mattrobball) [reply Mathlib/CategoryTheory/Limits/Shapes/Terminal.lean]: Elaborated a bit more.

**Review by jcommelin — APPROVED** [review]
Thanks :tada:

bors merge

**Pull request closed without merging**